### ISSUE TITLE
libheif AVIF API

### ISSUE THREAD

**Issue by YakoYakoYokuYoku** [libgd/libgd#688]
As HEIF support got included in #670, I propose to also include a `libheif` backend for AVIF.

Similar to #663.

**Comment by vapier** [comment]
we can do some compile-time logic at least.  like update src/gd_avif.c to, if `!HAVE_LIBAVIF && HAVE_LIBHEIF`, we have the Avif funcs call the Heif funcs.  wdyt ?

**Comment by YakoYakoYokuYoku** [comment]
My original plan was to do a `with-avif=<backend>` switch but we could go the preprocessor only way, although this does not satisfy the corner case of having both libs but only using `libheif`.

**Comment by vapier** [comment]
iiuc, code could explicitly call the Heif APIs on avif files right ?

**Comment by YakoYakoYokuYoku** [comment]
Yes, the `gd` AVIF implementation could call the existing HEIF methods. Now we have something like:

``` php
$avif = imagecreatefromheif("test_a.avif");
imageheif($image, "test_b.avif", 70, GD_HEIF_CODEC_AV1);
```

> My original plan was to do a ...

There I was speaking about the build system. Forgot to mention it 🤡.

**Comment by vapier** [comment]
if the goal is to make it easy for people writing php code to utilize libheif, i think doing this at runtime is best.  i expect most php users won't actually have the ability to build php (or gd) from source, they'll just be using whatever binaries are provided to them via their hosting services or distros.  so a build-time option wouldn't be super useful to them.

having the fallback with gd_avif.c calling into gd_heif.c would be helpful for the cases where people are building gd and have access to libheif but not libavif.  i think that would obviate the need for a dedicated `--with-xxx` configure flag as people could just `--without-avif` and it'd be provided by libheif automatically.

that said, using heif APIs to manipulate avif files when there are explicit avif APIs can be confusing to users.  we could call it out in the documentation to start with.  or some new API that would set preferred backends that you'd run once at the start of a script ?  not sure what form this would take.

**Comment by morsssss** [comment]
On the PHP front - PHP comes with [its own bundled gd fork](https://github.com/php/php-src/tree/master/ext/gd). Please ignore this comment if it's redundant information, but I didn't know it when I first got involved in creating AVIF support :)

[This page](https://www.php.net/manual/en/image.installation.php) describes the options - no GD support, support from the built-in GD fork, and support from the proper libgd.

**Comment by YakoYakoYokuYoku** [comment]
> if the goal is to make it easy for people writing php code to utilize libheif, i think doing this at runtime is best.

This and every other language, but more specifically the direct use of the C library.

> we could call it out in the documentation to start with.

I think docs should be the way.

**Comment by vapier** [comment]
@YakoYakoYokuYoku want to do this up ?  we'd want doc improvements, and gd_avif.c to fallback to libheif if libavif isn't available.

**Comment by YakoYakoYokuYoku** [comment]
Sure, I'll submit a PR.

**Comment by pierrejoye** [comment]
My apoligize, I did not dig yet into each format's specs.  would it be possible to have a bit more info about it?

I could not find any referene of HEIC in libavif docs but references in the src comments :)

To my understanding AVIF support sequences of images. HEIC not. Correct?

libAVIF can encode and decode HEIC image, Correct?

libheif cannot read avif images, correct?

As of now we don't support sequences, so using one or the other library won't be too confusing. However the moment we do, some functions available when libavif is used won't be available if libheif is used as "backend".

You get the idea. I would like to see what is possible by which. As of now, I feel like one may be used to read the other image formats but it is tricky to make it nice from an enduser point of view. I feel like it is more a job for the binding, or the application using GD (implementing its own fallback to the HEIF or AVIF GD APIs.

What do you think?
